src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+		if (routerQuery == NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Select query cannot be pushed down to the worker.")));
+		}
+
+		/* Ensure that we have INSERTed table's placement exists on the same worker */
+		insertShardPlacementList = ShardPlacementList(shardId);
+		intersectedPlacementList = IntersectPlacementList(insertShardPlacementList,
+														  selectPlacementList);
+
+		if (list_length(insertShardPlacementList) != list_length(
+				intersectedPlacementList))
+		{
+			ereport(DEBUG2, (errmsg("skipping the task"),
+							 errdetail("Insert query hits %d placements, Select query "
+									   "hits %d placements and only %d of those placements match.",
+									   list_length(insertShardPlacementList), list_length(
+										   selectPlacementList),
+									   list_length(intersectedPlacementList))));
+
+			continue;
+		}
+
+		/* this is required for correct deparsing of the query */
+		ReorderInsertSelectTargetListsIfExists(copiedOriginal);
+
+		/* setting an alias simplifies deparsing of RETURNING */
+		rangeTableEntry = linitial(copiedOriginal->rtable);
+		if (rangeTableEntry->alias == NULL)
+		{
+			Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+			rangeTableEntry->alias = alias;
+		}
+
+		/* and generate the full query string */
+		deparse_shard_query(copiedOriginal, distributedTableId, shardInterval->shardId,
+							queryString);
+		ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+
+		sqlTask = CreateBasicTask(jobId, taskIdIndex++, MODIFY_TASK, queryString->data);
+		sqlTask->dependedTaskList = NULL;
+		sqlTask->anchorShardId = shardId;
+		sqlTask->taskPlacementList = insertShardPlacementList;
+
+		sqlTaskList = lappend(sqlTaskList, sqlTask);
+	}
+
+
+	/* Create the worker job */
+	workerJob = CitusMakeNode(Job);
+	workerJob->taskList = sqlTaskList;
+	workerJob->subqueryPushdown = false;
+	workerJob->dependedJobList = NIL;
+	workerJob->jobId = jobId;
+	workerJob->jobQuery = originalQuery;
+	workerJob->requiresMasterEvaluation = false; /* for now we do not support any function evaluation */
+
+	/* and finally the multi plan */
+	multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->workerJob = workerJob;
+	multiPlan->masterTableName = NULL;
+	multiPlan->masterQuery = NULL;
+
+	return multiPlan;
+}
+
+
+/*
+ * ErrorIfInsertSelectQueryNotSupported errors out for unsupported
+ * INSERT ... SELECT queries.
+ */
+static void
+ErrorIfInsertSelectQueryNotSupported(Query *queryTree)
+{
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	/* we only do this check for INSERT ... SELECT queries */
+	AssertArg(InsertSelectQuery(queryTree));
+
+	subqueryRte = lsecond(queryTree->rtable);
+	subquery = subqueryRte->subquery;
+
+	/* we support this feature only for colocated tables */
+	ErrorIfNotAllParticipatingTablesAreColocated(queryTree);
```

Comment:
I think we simply should remove that check. We don't have it for selects either. And it'll break reference tables.